### ISSUE TITLE
[apex] ApexSOQLInjection allows combining a list where it shouldn't

### ISSUE THREAD

**Issue by gwilymatgearset** [pmd/pmd#4075]
**Affects PMD Version:** 6.47

**Rule:** [ApexSOQLInjection](https://pmd.github.io/latest/pmd_rules_apex_security.html#apexsoqlinjection)

**Description:**

@adangel spotted that PMD allows 'Safe SOQL with List concatenation', but it is too lenient (comment: https://github.com/pmd/pmd/pull/4056/files/94b3723da0f289ee275e4a94744c196d43b599f8#r929129197).

You are able to craft soql injections give the limitations allowed in the rule, assuming you have control of list elements.

**Code Sample demonstrating the issue:**

This code was run in the execute anonymous

```apex
List<string> stuff = new List<string>(); // assume stuff comes from a malicious source

stuff.add('WHERE Name = \'hi\' OR Name = \'');
stuff.add('\'');

Database.countQuery(
    'SELECT COUNT() FROM Account ' +
    string.join(stuff, ',')
);
```

**Expected outcome:**

PMD should report a violation at line 8, but doesn't. This is a false-negative.

**Running PMD through:** CLI
